source/extensions/clusters/redis/redis_cluster_lb.cc

```diff
@@ -55,17 +126,62 @@ RedisClusterLoadBalancer::chooseHost(Envoy::Upstream::LoadBalancerContext* conte
     return nullptr;
   }
 
-  return slot_array_->at(hash.value() % Envoy::Extensions::Clusters::Redis::MaxSlot);
+  auto shard = shard_vector_->at(
+      slot_array_->at(hash.value() % Envoy::Extensions::Clusters::Redis::MaxSlot));
+
+  auto redis_context = dynamic_cast<RedisLoadBalancerContext*>(context);
+  if (redis_context && redis_context->isReadCommand()) {
+    switch (redis_context->readPolicy()) {
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::Master:
+      return shard->master();
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::PreferMaster:
+      if (shard->master()->health() == Upstream::Host::Health::Healthy) {
+        return shard->master();
+      } else {
+        return chooseRandomHost(shard->allHosts(), random_);
+      }
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::Replica:
+      return chooseRandomHost(shard->replicas(), random_);
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::PreferReplica:
+      if (!shard->replicas().healthyHosts().empty()) {
+        return chooseRandomHost(shard->replicas(), random_);
+      } else {
+        return chooseRandomHost(shard->allHosts(), random_);
+      }
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::Any:
+      return chooseRandomHost(shard->allHosts(), random_);
+    default:
```

Comment:
Is this needed?

Reply:
This is not needed, but I thought this is what we do in the codebase to catch things in case another value got added in the future.

Reply:
The compiler won't compile the code with a missing switch case, you can remove it.

/wait